rectools/metrics/auc.py

```diff
@@ -0,0 +1,512 @@
+        recommendations in `recommend` method. For example, ItemKNN generates recommendations only
+        until the model has non-zero scores for the item in item-item similarity matrix. So with
+        small `K` for neighbours in ItemKNN and big `K` for `recommend` and AUC based metric you
+        will still get an error when `insufficient_handling` is set to `raise`.
+    """
+
+    def __init__(self, k: int, insufficient_handling: tp.Optional[str] = "ignore") -> None:
+        super().__init__(k=k)
+        try:
+            self.insufficient_handling = InsufficientHandling(insufficient_handling)
+        except ValueError:
+            possible_values = {item.value for item in InsufficientHandling.__members__.values()}
+            raise ValueError(
+                f"`insufficient_handling` must be one of the {possible_values}. Got {insufficient_handling}."
+            )
+
+    @classmethod
+    def fit(cls, reco: pd.DataFrame, interactions: pd.DataFrame, k_max: int) -> AUCFitted:
+        """
+        Prepare intermediate data for effective calculation.
+
+        You can use this method to prepare some intermediate data
+        for later calculation. It can optimize calculations if
+        you want calculate different AUC based metrics with different `k` parameter.
+        """
+        cls._check(reco, interactions=interactions)
+
+        outer_merged = outer_merge_reco(reco, interactions)
+        recommended_mask = ~outer_merged[Columns.Rank].isna()
+        outer_merged["__tp"] = recommended_mask & outer_merged["__test_positive"]
+        outer_merged["__fp"] = recommended_mask & ~outer_merged["__test_positive"]
+
+        stats = outer_merged.groupby(Columns.User).agg(
+            __fp_cumsum=("__fp", "cumsum"),
+            __test_pos_cumsum=("__test_positive", "cumsum"),
+            n_pos=("__test_positive", "sum"),
+            n_fp=("__fp", "sum"),
+        )
+        n_pos = stats["n_pos"].dropna().rename_axis(Columns.User)
+        users_n_fp = stats["n_fp"].dropna().rename_axis(Columns.User)
+        outer_merged = pd.concat([outer_merged, stats[["__fp_cumsum", "__test_pos_cumsum"]]], axis=1)
+
+        # Every user with FP count more then k_max has sufficient recommendations for partial AUC based metrics
+        # We calculate and keep number of false positives for all other users
+        n_fp_insufficient = users_n_fp[users_n_fp < k_max]
+        users_with_fn = outer_merged.loc[outer_merged[Columns.Rank].isna(), Columns.User].unique()
+        n_fp_insufficient = n_fp_insufficient[n_fp_insufficient.index.isin(users_with_fn)]
+
+        return AUCFitted(outer_merged, n_pos, n_fp_insufficient)
+
+    def _get_sufficient_reco_explanation(self) -> str:
+        raise NotImplementedError()
+
+    def _handle_insufficient_cases(
+        self, outer_merged: pd.DataFrame, n_pos: pd.Series, n_fp_insufficient: pd.Series, metric_name: str
```

Comment:
It's simpler, more correct and more stable to take the class name as a metric name

Reply:
fixed